src/core/core.layouts.js

```diff
@@ -49,9 +99,142 @@ function findMaxPadding(boxes) {
 }
 
 function addSizeByPosition(boxes, size) {
-	helpers.each(boxes, function(box) {
+	var i, ilen, box;
+	for (i = 0, ilen = boxes.length; i < ilen; ++i) {
+		box = boxes[i].box;
 		size[box.position] += box.isHorizontal() ? box.height : box.width;
-	});
+	}
+}
+
+function getMinSize(layout, chartSize) {
+	var box = layout.box;
+	var minSize;
+
+	if (layout.isHorizontal) {
+		minSize = box.update(box.fullWidth ? chartSize.width : chartSize.maxWidth, chartSize.horzMaxHeight);
+		chartSize.maxHeight -= minSize.height;
+	} else {
+		minSize = box.update(chartSize.vertMaxWidth, chartSize.maxHeight);
+		chartSize.maxWidth -= minSize.width;
+	}
+	return minSize;
+}
+
+function updateMinSizes(boxes, chartSize, outerBoxSizes) {
+	var i, ilen, layout;
+
+	for (i = 0, ilen = boxes.length; i < ilen; ++i) {
+		layout = boxes[i];
+
+		layout.isHorizontal = layout.box.isHorizontal();
+		layout.minSize = getMinSize(layout, chartSize);
+	}
+	addSizeByPosition(boxes, outerBoxSizes);
+}
+
+function fitHorizontalBoxes(boxes, chartSize, outerBoxSizes, maxPadding) {
+	var i, ilen, layout, box, margin;
+
+	for (i = 0, ilen = boxes.length; i < ilen; ++i) {
+		layout = boxes[i];
+		box = layout.box;
+
+		margin = {
+			left: Math.max(outerBoxSizes.left, maxPadding.left),
+			right: Math.max(outerBoxSizes.right, maxPadding.right),
+			top: 0,
+			bottom: 0
+		};
+
+		// Don't use min size here because of label rotation. When the labels are rotated, their rotation highly depends
+		// on the margin. Sometimes they need to increase in size slightly
+		box.update(box.fullWidth ? chartSize.width : chartSize.maxWidth, chartSize.horzMaxHeight, margin);
+	}
+	addSizeByPosition(boxes, outerBoxSizes);
+}
+
+function fitVerticalBoxes(boxes, chartSize, outerBoxSizes, maxPadding) {
+	var i, ilen, layout, box, margin;
+
+	for (i = 0, ilen = boxes.length; i < ilen; ++i) {
+		layout = boxes[i];
+		box = layout.box;
+
+		margin = boxes.fitted && {
+			left: 0,
+			right: 0,
+			top: Math.max(outerBoxSizes.top, maxPadding.top),
+			bottom: Math.max(outerBoxSizes.bottom, maxPadding.bottom)
+		};
+
+		box.update(chartSize.vertMaxWidth, chartSize.maxHeight, margin);
+	}
+	if (!boxes.fitted) {
+		addSizeByPosition(boxes, outerBoxSizes);
+	}
+	boxes.fitted = true;
```

Comment:
I don't love that this method has a side-effect of altering the `boxes` input. What if we instead checked if `maxPadding` was present? (I just left another comment that we should pass it in only on the 2nd time)

Reply:
I changed the parameters a bit. but could add another parameter now to indicate final fit.

Reply:
Noticed couple of leftover extra parameters passed and added `finalFit`.